File: frontend/src/api/metrics.ts
import client from "./client";

export interface OpsPerSecond {
  insert: number | null;
  query: number | null;
  update: number | null;
  delete: number | null;
}

export interface MetricLatest {
  cluster_id: string;
  collected_at: string;
  rs_state: string;
  primary_member: string | null;
  members_up: number;
  members_total: number;
  max_replication_lag_seconds: number | null;
  connections: { current: number | null; available: number | null };
  ops_per_second: OpsPerSecond;
  memory: { resident_mb: number | null; virtual_mb: number | null };
  cache: { used_bytes: number | null; total_bytes: number | null; dirty_bytes: number | null; hit_ratio: number | null };
  storage: { data_size_bytes: number | null; fs_total_bytes: number | null; fs_used_bytes: number | null; fs_used_percent: number | null };
  status: string;
}

export interface MetricPoint {
  collected_at: string;
  connections_current: number | null;
  max_replication_lag_seconds: number | null;
  memory_resident_mb: number | null;
  fs_used_percent: number | null;
  ops_per_second: OpsPerSecond;
}

export interface MetricRange {
  cluster_id: string;
  range: string;
  points: MetricPoint[];
}

export async function fetchLatestMetrics(clusterId: string): Promise<MetricLatest> {
  return (await client.get<MetricLatest>(`/clusters/${clusterId}/metrics/latest`)).data;
}

export async function fetchMetricsRange(clusterId: string, range: string): Promise<MetricRange> {
  return (await client.get<MetricRange>(`/clusters/${clusterId}/metrics`, { params: { range } })).data;
}

export async function refreshMetrics(clusterId: string): Promise<MetricLatest> {
  return (await client.post<MetricLatest>(`/clusters/${clusterId}/metrics/refresh`)).data;
}

File: backend/app/api/schemas/metrics.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class ConnectionsInfo(BaseModel):
    current: int | None = None
    available: int | None = None


class OpsPerSecond(BaseModel):
    insert: float | None = None
    query: float | None = None
    update: float | None = None
    delete: float | None = None


class MemoryInfo(BaseModel):
    resident_mb: int | None = None
    virtual_mb: int | None = None


class CacheInfo(BaseModel):
    used_bytes: int | None = None
    total_bytes: int | None = None
    dirty_bytes: int | None = None
    hit_ratio: float | None = None


class StorageInfo(BaseModel):
    data_size_bytes: int | None = None
    fs_total_bytes: int | None = None
    fs_used_bytes: int | None = None
    fs_used_percent: float | None = None


class MetricLatestResponse(BaseModel):
    cluster_id: str
    collected_at: datetime
    rs_state: str
    primary_member: str | None = None
    members_up: int
    members_total: int
    max_replication_lag_seconds: float | None = None
    connections: ConnectionsInfo
    ops_per_second: OpsPerSecond
    memory: MemoryInfo
    cache: CacheInfo
    storage: StorageInfo
    status: str


class MetricPointResponse(BaseModel):
    collected_at: datetime
    connections_current: int | None = None
    max_replication_lag_seconds: float | None = None
    memory_resident_mb: int | None = None
    fs_used_percent: float | None = None
    ops_per_second: OpsPerSecond


class MetricRangeResponse(BaseModel):
    cluster_id: str
    range: str
    points: list[MetricPointResponse]
